// @webflow/webflow-cli/dist/cloud-scaffolds/fetcher.ts
import * as fs from "fs";
import fsExtra from "fs-extra";
import * as path from "path";
import * as os from "os";
import * as tar from "tar";
import { getFrameworkDefinition, toMinimalScaffoldId } from "./registry";
import type { ScaffoldContent } from "./types";
import { ScaffoldError } from "./types";
import logger from "../../utils/logger";

const GITHUB_TARBALL_BASE = "https://github.com";

/**
 * Get scaffold content for a framework.
 * Resolves the scaffold ID (e.g. "astro" -> "astro-minimal") and fetches.
 *
 * @param frameworkOrScaffoldId - "astro", "nextjs", or full "astro-minimal", "nextjs-minimal"
 * @returns ScaffoldContent ready for applyScaffold
 */
export async function getScaffoldContent(
  frameworkOrScaffoldId: string
): Promise<ScaffoldContent> {
  const scaffoldId = frameworkOrScaffoldId.endsWith("-minimal")
    ? frameworkOrScaffoldId
    : toMinimalScaffoldId(frameworkOrScaffoldId);
  return fetchScaffoldContent(scaffoldId);
}

/**
 * Get scaffold content from the configured source (github or local).
 * Dispatches to the appropriate fetcher based on source.type.
 */
export async function fetchScaffoldContent(
  scaffoldId: string
): Promise<ScaffoldContent> {
  const def = getFrameworkDefinition(scaffoldId);

  if (def.source.type === "github") {
    return fetchScaffoldFromGitHub(scaffoldId);
  }
  return getScaffoldFromLocal(scaffoldId);
}

/**
 * Get scaffold content from local filesystem.
 */
export async function getScaffoldFromLocal(
  scaffoldId: string
): Promise<ScaffoldContent> {
  const def = getFrameworkDefinition(scaffoldId);

  if (def.source.type !== "local") {
    throw new ScaffoldError(
      `Scaffold "${scaffoldId}" does not have a local source`,
      scaffoldId
    );
  }

  // path is relative to dist/ (e.g. "cloud-scaffolds/astro-minimal") or absolute
  const scaffoldPath = path.isAbsolute(def.source.path)
    ? def.source.path
    : path.join(__dirname, def.source.path);
  if (!fs.existsSync(scaffoldPath)) {
    throw new ScaffoldError(
      `Scaffold directory not found: ${scaffoldPath}`,
      scaffoldId
    );
  }

  const files = await listFilesRecursive(scaffoldPath);
  return { scaffoldId, files, extractionRoot: scaffoldPath };
}

/**
 * Download and extract a scaffold from GitHub.
 * Uses the tarball API; no git dependency.
 */
async function fetchScaffoldFromGitHub(
  scaffoldId: string
): Promise<ScaffoldContent> {
  const def = getFrameworkDefinition(scaffoldId);

  if (def.source.type !== "github") {
    throw new ScaffoldError(
      `Scaffold "${scaffoldId}" does not have a GitHub source`,
      scaffoldId
    );
  }

  const ref = def.source.ref ?? "main";
  const tarballUrl = `${GITHUB_TARBALL_BASE}/${def.source.repo}/archive/${ref}.tar.gz`;

  const extractDir = await fs.promises.mkdtemp(
    path.join(os.tmpdir(), `webflow-scaffold-${scaffoldId}-`)
  );

  const downloadSpinner = logger.spinner(
    `Downloading scaffold (${def.source.repo}@${ref})`
  );

  try {
    let response: Response;
    try {
      response = await fetch(tarballUrl, {
        redirect: "follow",
        headers: { "User-Agent": "webflow-cli" },
      });
    } catch (fetchErr) {
      downloadSpinner.fail("Download failed");
      throw new ScaffoldError(
        `Failed to fetch scaffold from GitHub: ${
          fetchErr instanceof Error ? fetchErr.message : String(fetchErr)
        }. Repo: ${def.source.repo}, ref: ${ref}`,
        scaffoldId
      );
    }

    if (!response.ok) {
      downloadSpinner.fail(
        `GitHub returned ${response.status} ${response.statusText}`
      );
      throw new ScaffoldError(
        `Failed to fetch scaffold from GitHub: ${response.status} ${response.statusText}. ` +
          `Repo: ${def.source.repo}, ref: ${ref}`,
        scaffoldId
      );
    }

    const buffer = Buffer.from(await response.arrayBuffer());
    downloadSpinner.succeed("Downloaded scaffold");

    const unpackSpinner = logger.spinner("Unpacking scaffold");
    try {
      await extractTarball(buffer, extractDir);
      unpackSpinner.succeed("Scaffold ready");
    } catch (unpackErr) {
      unpackSpinner.fail("Unpack failed");
      throw unpackErr;
    }

    const entries = await fs.promises.readdir(extractDir);
    const topLevelDir = entries[0];
    if (!topLevelDir || entries.length !== 1) {
      throw new ScaffoldError(
        `Unexpected tarball structure for ${def.source.repo}`,
        scaffoldId
      );
    }

    const extractionRoot = path.join(extractDir, topLevelDir);
    const files = await listFilesRecursive(extractionRoot);

    return {
      scaffoldId,
      files,
      extractionRoot,
    };
  } catch (err) {
    try {
      fsExtra.removeSync(extractDir);
    } catch {
      // Ignore cleanup errors
    }
    if (err instanceof ScaffoldError) throw err;
    throw new ScaffoldError(
      `Failed to fetch scaffold "${scaffoldId}": ${
        err instanceof Error ? err.message : String(err)
      }`,
      scaffoldId
    );
  }
}

async function extractTarball(buffer: Buffer, destDir: string): Promise<void> {
  const tmpFile = path.join(
    os.tmpdir(),
    `webflow-scaffold-${Date.now()}.tar.gz`
  );
  await fs.promises.writeFile(
    tmpFile,
    new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength)
  );
  try {
    await tar.extract({ file: tmpFile, cwd: destDir });
  } finally {
    await fs.promises.unlink(tmpFile).catch(() => {
      // Ignore cleanup errors
    });
  }
}

async function listFilesRecursive(dirPath: string): Promise<string[]> {
  const result: string[] = [];
  const entries = await fs.promises.readdir(dirPath, { withFileTypes: true });

  for (const entry of entries) {
    const fullPath = path.join(dirPath, entry.name);
    if (entry.isDirectory()) {
      result.push(...(await listFilesRecursive(fullPath)));
    } else {
      result.push(fullPath);
    }
  }
  return result;
}
