// @webflow/webflow-cli/dist/cloud-scaffolds/types.ts
/**
 * GitHub-based scaffold source.
 * Used when fetching from a public GitHub repository.
 */
export interface GitHubScaffoldSource {
  type: "github";
  /** Repository in "owner/repo" format */
  repo: string;
  /** Branch or tag. Default: "main" */
  ref?: string;
}

/**
 * Local scaffold source.
 * Used when reading from bundled scaffold directories.
 */
export interface LocalScaffoldSource {
  type: "local";
  /** Path to scaffold directory (relative to dist root or absolute) */
  path: string;
}

/**
 * Union of supported scaffold sources.
 * Each variant has only the properties it needs — no optional repo on local.
 * The `type` discriminant enables TypeScript narrowing.
 */
export type ScaffoldSource = GitHubScaffoldSource | LocalScaffoldSource;

/**
 * Definition of a supported framework scaffold.
 * Single entry point for adding new frameworks.
 */
export interface FrameworkDefinition {
  /** Unique ID used in CLI (e.g. "astro-minimal", "nextjs-minimal") */
  id: string;
  /** Display name for prompts */
  name: string;
  /** Where to fetch the scaffold content */
  source: ScaffoldSource;
}

/**
 * Result of fetching scaffold content.
 * Used by applyScaffold to write files.
 */
export interface ScaffoldContent {
  /** Scaffold identifier */
  scaffoldId: string;
  /** Absolute paths to scaffold files */
  files: string[];
  /** Root directory of the scaffold (for path resolution) */
  extractionRoot: string;
}

/**
 * Error thrown when a scaffold is not found or unsupported.
 */
export class ScaffoldError extends Error {
  constructor(
    message: string,
    public readonly scaffoldId?: string,
    public readonly supportedIds?: string[]
  ) {
    super(message);
    this.name = "ScaffoldError";
  }
}
