// @webflow/webflow-cli/dist/templates/vite/vite.config.webflow.ts
// Merge logic for the generated vite.config.ts. Copied verbatim into the
// user's project; types are local to avoid Vite cross-version drift.

type ViteConfigEnv = { mode: string; command: string; isPreview?: boolean };
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ViteUserConfig = Record<string, any>;
type ViteUserConfigFn = (
  env: ViteConfigEnv
) => ViteUserConfig | Promise<ViteUserConfig>;

export type UserViteConfig = ViteUserConfig | ViteUserConfigFn;
export type WebflowOverridesOptions = {
  basePath: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  cloudflarePlugins: any[];
};

/**
 * Wrap the user's Vite config with Webflow Cloud's required overrides.
 * `userExport` is always provided — `{}` when the user ships no vite.config.*.
 *
 * Returns an async factory so the resulting config plays nicely with Vite's
 * loadConfigFromFile under both ESM and CJS package types.
 */
export function withWebflowOverrides(
  userExport: UserViteConfig,
  opts: WebflowOverridesOptions
): ViteUserConfigFn {
  return async function viteConfig(
    env: ViteConfigEnv
  ): Promise<ViteUserConfig> {
    const resolved = await resolveUserConfig(userExport, env);
    return {
      ...resolved,
      base: opts.basePath,
      define: {
        ...resolved.define,
        __WF_BASE_PATH__: JSON.stringify(opts.basePath),
        __WF_BASE_URL__: JSON.stringify(
          opts.basePath.endsWith("/") ? opts.basePath : opts.basePath + "/"
        ),
      },
      // Cloudflare's plugins go first so they sit at the head of the chain;
      // user plugins are preserved after them (matching pre-1.2 behaviour).
      plugins: [...opts.cloudflarePlugins, ...(resolved.plugins || [])],
    };
  };
}

async function resolveUserConfig(
  userExport: UserViteConfig,
  env: ViteConfigEnv
): Promise<ViteUserConfig> {
  if (typeof userExport === "function") {
    return (await userExport(env)) ?? {};
  }
  return userExport ?? {};
}
